torch/utils/data/_utils/worker.py

```diff
@@ -87,23 +149,34 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
                 continue
             if r is None:
                 # Received the final signal
-                assert done_event.is_set()
+                assert done_event.is_set() or iteration_end
                 return
-            elif done_event.is_set():
+            elif done_event.is_set() or iteration_end:
                 # Done event is set. But I haven't received the final signal
                 # (None) yet. I will keep continuing until get it, and skip the
                 # processing steps.
                 continue
-            idx, batch_indices = r
+            idx, index = r
             try:
-                samples = collate_fn([dataset[i] for i in batch_indices])
+                if mode == DataLoaderMode.Iterable:
+                    try:
+                        data = convert_fn(next(dataset_iter))
+                    except StopIteration:
+                        data = IterableDatasetStopIteration(worker_id)
+                        iteration_end = True  # don't waste resource in future iter
```

Comment:
`try..except..`. The other important use of this flag is to prevent sending more than one `IterableDatasetStopIteration` to main proc. I will add this in the comment.